salt/modules/purefa.py

```diff
@@ -0,0 +1,1274 @@
+             }
+
+    user_agent = '{base} {class}/{version} ({platform})'.format(**agent)
+
+    try:
+        array = __opts__['pure_tags']['fa'].get('san_ip')
+        api = __opts__['pure_tags']['fa'].get('api_token')
+        if array and api:
+            system = purestorage.FlashArray(array, api_token=api, user_agent=user_agent)
+    except (KeyError, NameError, TypeError):
+        try:
+            san_ip = os.environ.get('PUREFA_IP')
+            api_token = os.environ.get('PUREFA_API')
+            system = purestorage.FlashArray(san_ip,
+                                            api_token=api_token,
+                                            user_agent=user_agent)
+        except (ValueError, KeyError, NameError):
+            try:
+                system = purestorage.FlashArray(__pillar__['PUREFA_IP'],
+                                                api_token=__pillar__['PUREFA_API'],
+                                                user_agent=user_agent)
+            except (KeyError, NameError):
+                raise CommandExecutionError('No Pure Storage FlashArray credentials found.')
+
+    try:
+        system.get()
+    except Exception:
+        raise CommandExecutionError('Pure Storage FlashArray authentication failed.')
+    return system
+
+
+def _get_volume(name, array):
+    '''Private function to check volume'''
+    try:
+        return array.get_volume(name)
+    except purestorage.PureError:
+        return None
+
+
+def _get_snapshot(name, suffix, array):
+    '''Private function to check snapshot'''
+    snapshot = name + '.' + suffix
+    try:
+        for snap in array.get_volume(name, snap=True):
+            if snap['name'] == snapshot:
+                return snapshot
+    except purestorage.PureError:
+        return None
+
+
+def _get_deleted_volume(name, array):
+    '''Private function to check deleted volume'''
+    try:
+        return array.get_volume(name, pending='true')
+    except purestorage.PureError:
+        return None
+
+
+def _get_pgroup(name, array):
+    '''Private function to check protection group'''
+    pgroup = None
+    for temp in array.list_pgroups():
+        if temp['name'] == name:
+            pgroup = temp
+            break
+    return pgroup
+
+
+def _get_deleted_pgroup(name, array):
+    '''Private function to check deleted protection group'''
+    try:
+        return array.get_pgroup(name, pending='true')
+    except purestorage.PureError:
+        return None
+
+
+def _get_hgroup(name, array):
+    '''Private function to check hostgroup'''
+    hostgroup = None
+    for temp in array.list_hgroups():
+        if temp['name'] == name:
+            hostgroup = temp
+            break
+    return hostgroup
+
+
+def _get_host(name, array):
+    '''Private function to check host'''
+    host = None
+    for temp in array.list_hosts():
+        if temp['name'] == name:
+            host = temp
+            break
+    return host
+
+
+def _human_to_bytes(size):
```

Comment:
This would be a good candidate to move into Salt's `utils` functions if you would like to move it there and then import it into this. Many other modules could then benefit.

Reply:
@cachedout - do you have a specific existing file you would like this and the _xor function to be added to?

Reply:
Please go ahead and move it to `__init__.py` in `salt.utils`. Thanks!